### PR TITLE
Cleanup test_count_matched test to handle non-DDS RMWs

### PR DESCRIPTION
**Pull request by clalancette**
The main goal of this PR is to update the test_count_matched test so that it handles non-DDS RMWs.  Along the way, it also does some cleanup in the test.  The patches are:

1.  Slightly refactor things so that `check_state` is a protected class method, rather than a free function.  That way it can use more of the internal details of the class, which it depends on anyway.
2.  Renames "ops" to "opts", as that better reflects that these are options.
3.  Uses scope_exit to do some cleanup in the second test.  We purposefully do *not* do this in the first test, as it specifically tests what happens as various things get destroyed.
4.  We add compatibility with non-DDS RMWs.  Other RMWs may have different matching semantics between different QoS settings.  In particular, `rmw_zenoh` will happily match BEST_EFFORT publishers to RELIABLE subscriptions, while the DDS RMWs will not.  Luckily, we have the function called `rmw_qos_profile_check_compatible` that we can call to tell whether the underlying RMW would match them, so we can use that here.

### CODE REVIEW

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21338/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15650/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/852/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22076/)

File: rcl/test/rcl/test_count_matched.cpp [hunk 1]
```diff
@@ -141,47 +83,109 @@ class TestCountFixture : public ::testing::Test
...
+
+protected:
+  const rcl_guard_condition_t * graph_guard_condition;
+  rcl_context_t * context_ptr;
+  rcl_wait_set_t * wait_set_ptr;
+
+  void check_state(
+    rcl_publisher_t * publisher,
+    rcl_subscription_t * subscriber,
+    size_t expected_subscriber_count,
```
Comment (ahcorde) [inline rcl/test/rcl/test_count_matched.cpp:97]: include `<cstddef>`

File: rcl/test/rcl/test_count_matched.cpp [hunk 2]
```diff
@@ -141,47 +83,109 @@ class TestCountFixture : public ::testing::Test
...
+      }
+
+      ret = rcl_wait_set_clear(wait_set_ptr);
+      ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+      ret = rcl_wait_set_add_guard_condition(wait_set_ptr, graph_guard_condition, NULL);
+      ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+      std::chrono::nanoseconds time_to_sleep = std::chrono::milliseconds(200);
+      RCUTILS_LOG_INFO_NAMED(
+        ROS_PACKAGE_NAME,
+        "  state wrong, waiting up to '%s' nanoseconds for graph changes... ",
```
Comment (ahcorde) [inline rcl/test/rcl/test_count_matched.cpp]:
```suggestion
        " state wrong, waiting up to '%s' nanoseconds for graph changes... ",
```

**Review by ahcorde — APPROVED** [review]
Two minor comments, LGTM with green CI

File: rcl/test/rcl/test_count_matched.cpp (hunk 1 above)
Comment (clalancette) [reply rcl/test/rcl/test_count_matched.cpp:97]: Good call, fixed in 8d468ca.

File: rcl/test/rcl/test_count_matched.cpp (hunk 2 above)
Comment (clalancette) [reply rcl/test/rcl/test_count_matched.cpp]: Actually, we don't need the space at all.  `RCUTILS_LOG_INFO_NAMED` will automatically add a space.  So I removed both spaces, and also removed it in another call above in 8d468ca.

**Comment by clalancette** [comment]
New CI after changes from review:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21339/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15652/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/853/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22077/)

**Review by fujitatomoya — APPROVED** [review]
lgtm with green CI

**Review by Yadunund — APPROVED** [review]

**Pull request merged**